### PR TITLE
[4-26] Collecting data from test

### PR DESCRIPTION
**Pull request by GeranSkot**
1. Added spinners to all modules on the test page. (6-32 task)
1.1. If there are no questions, there will be a redirect to the error page.
2. Integrated user-assigned tests. (6-18 task)
2.1. During the demonstration, there was a test, but at the moment an empty array comes in, I cannot demonstrate.
myAssignedTests
3. Created submit-test service, to collect data from test modules,
generate test and send answers to server. (6-22 task)
testServicei
4. Fixed bug in grammar/audition module. (6-23 task)
4.1. Don't remember exactly what I'm done.
5. Fixed bug answered the question in grammar/audition module. (6-24 task)
5.1. First grammar
answered1
5.2. Second audition
answered2
6. Added track test expiration. (6-25 task)
timerEndModal
7. Created modal dialog after finished the test. (6-26 task)
finishTestModal
8. Refactored timer component. (6-28 task)
timerService
9. Start content user name and level(KuznetsovR made the same:)) (6-38 task)
main-content

### CODE REVIEW

**Review by KotdenGit — APPROVED** [review]

File: exalearn-ui/src/app/components/data-table/data-table.component.html [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
 		<ng-container matColumnDef="assigned by">
 			<th mat-header-cell *matHeaderCellDef mat-sort-header>{{ 'TABLES.ASSIGNED_BY' | translate}}</th>
-			<td mat-cell *matCellDef="let element">{{element.assigner}}</td>
+			<td mat-cell *matCellDef="let element">{{element.assigner || element.assignedBy}}</td>
```
Comment (vstrabko) [inline exalearn-ui/src/app/components/data-table/data-table.component.html:50]: Why two different keys here?
Is it expected?

File: exalearn-ui/src/app/components/data-table/data-table.component.ts [hunk 2]
```diff
@@ -18,12 +18,13 @@ export class DataTableComponent implements AfterViewInit, OnInit {
 	dataSource: MatTableDataSource<PassedTest | UserBack>;
 
 	@Input() displayedColumns: string[];
-	@Input() data: object[];
-	@Input() displaySearch: boolean = true;
+	@Input() data: any;
```
Comment (vstrabko) [inline exalearn-ui/src/app/components/data-table/data-table.component.ts:21]: Should this really be "any"?
How come we don't know the type?

File: exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss [hunk 3]
```diff
@@ -9,20 +9,21 @@
 	width: 100%;
 }
 
+.mat-dialog-content {
+	overflow: visible;
+	display: flex;
+	flex-direction: row !important;
+	justify-content: space-evenly !important;
```
Comment (vstrabko) [inline exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss]: I can see that it was used without `!important` before.
Can we avoid using `!important`?

File: exalearn-ui/src/app/services/audio-cloud.service.ts [hunk 4]
```diff
@@ -7,29 +7,20 @@ import { finalize } from 'rxjs/operators';
...
-		}
-	];
-
-	getFiles() {
-		return of(this.files);
-	}
-
 	private storageRef: AngularFireStorageReference;
 	private uploadFile: AngularFireUploadTask;
+	files: any;
```
Comment (vstrabko) [inline exalearn-ui/src/app/services/audio-cloud.service.ts]: Do we really not now which type is file going to be?

File: exalearn-ui/src/app/services/submit-test.service.ts [hunk 5]
```diff
@@ -1,9 +1,46 @@
...
 
 @Injectable({
 	providedIn: 'root'
 })
 export default class SubmitTestService {
-	constructor() {}
+	constructor(private apiService: ApiService, private router: Router) {}
+
+	public mapOfAnswers = new Map();
+	public test: any;
```
Comment (vstrabko) [inline exalearn-ui/src/app/services/submit-test.service.ts]: Same. Why type `any` here?

File: exalearn-ui/src/app/services/timer.service.ts [hunk 6]
```diff
@@ -12,39 +16,74 @@ export class TimerService {
 	public testTotalTimer: number = TEST_DURATION;
 	public speakingTimer = 0;
 	public time: { mins: string; secs: string } = { mins: '00', secs: '00' };
+	public speakingTimerInterval: any;
+	public timerInterval: any;
+	timeForTimer: any;
+	timeForRecord: any;
```
Comment (vstrabko) [inline exalearn-ui/src/app/services/timer.service.ts]: Same Same. Why type any here?

File: exalearn-ui/src/app/test-data/test-questions.ts [hunk 7]
```diff
@@ -4,23 +4,145 @@ import { Question } from '../interfaces/interfaces';
...
 	{
-		id: 1123,
-		questionText: 'He got too tired _____ over work.',
-		choices: ['because of', 'because off', 'on', 'for'],
+		id: 1,
+		question: '1',
+		answers: ['5', '6', '7', '8'],
 		userAnswer: null,
 		index: undefined
 	},
```
Comment (vstrabko) [inline exalearn-ui/src/app/test-data/test-questions.ts:19]: Do we still need this mock data?

**Review by vstrabko — CHANGES_REQUESTED** [review]
Looks real nice!
Good job fixing code styling (deleting not necessary spaces, inserting new lines and etc.)
Some comments about use of `!important` and types in several methods and variables.

File: exalearn-ui/src/app/test-data/test-questions.ts (hunk 7 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/test-data/test-questions.ts:19]: Yes, we need this data for Yura V

File: exalearn-ui/src/app/components/data-table/data-table.component.html (hunk 1 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/data-table/data-table.component.html:50]: Added TODO

File: exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss (hunk 3 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss]: Refactored, ready

File: exalearn-ui/src/app/services/timer.service.ts (hunk 6 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/services/timer.service.ts]: Refactored, thx

File: exalearn-ui/src/app/components/audition/audition.component.scss [hunk 8]
```diff
@@ -22,6 +22,10 @@
 	align-items: baseline;
 }
 
+.spinner {
+	margin: 25% auto 25% auto;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/audition/audition.component.scss]: Can be replaced by `margin: 25% auto`

File: exalearn-ui/src/app/services/submit-test.service.ts (hunk 5 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/services/submit-test.service.ts]: Refactored, created new interface "Test"

File: exalearn-ui/src/app/components/audition/audition.component.ts [hunk 9]
```diff
@@ -9,9 +9,22 @@ import { Question } from '../../interfaces/interfaces';
 	encapsulation: ViewEncapsulation.None
 })
 export class AuditionComponent implements OnInit {
+	@Input() questionsAudio: Question[];
+	@Input() testPassedId: number;
+
 	testQuestions: Question[] = [];
+	audioUrl: string;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/audition/audition.component.ts]: Just why one field has `public` and others no?)

File: exalearn-ui/src/app/services/audio-cloud.service.ts (hunk 4 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/services/audio-cloud.service.ts]: Created new type, thx

File: exalearn-ui/src/app/components/essay-part/essay-part.component.scss [hunk 10]
```diff
@@ -25,3 +26,13 @@
 	min-height: 200px;
 	resize: none;
 }
+
+.spinner {
+	margin: 25% auto 25% auto;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/essay-part/essay-part.component.scss]: Same as above

File: exalearn-ui/src/app/components/essay-part/essay-part.component.scss [hunk 11]
```diff
@@ -25,3 +26,13 @@
...
+}
+
+@media screen and (max-width: 650px) {
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/essay-part/essay-part.component.scss:34]: Can we use our mediaQuery mixins?

File: exalearn-ui/src/app/components/finish-test-modal/finish-test-modal.component.scss [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+.modal-parameter {
+	width: 100%;
+}
+
+.mat-dialog-content {
+	overflow: visible;
+	display: flex;
+	flex-direction: row !important;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/finish-test-modal/finish-test-modal.component.scss]: Do we really need ! important here?

File: exalearn-ui/src/app/components/finish-test-modal/finish-test-modal.component.scss [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+	justify-content: space-evenly !important;
+}
+
+::ng-deep {
+	.mat-dialog-container {
+		text-align: center;
+	}
+
+	.mat-raised-button {
+		padding: 0 10px 0 10px;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/finish-test-modal/finish-test-modal.component.scss]: Same as above

File: exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss [hunk 14]
```diff
@@ -9,20 +9,21 @@
...
 
-	.mat-dialog-content {
-		overflow: visible;
-		display: flex;
-		flex-direction: row;
-		justify-content: space-evenly;
-	}
-
 	.mat-raised-button {
 		padding: 0 10px 0 10px;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss]: Same as above

File: exalearn-ui/src/app/components/question/question.component.scss [hunk 15]
```diff
@@ -21,9 +22,12 @@
 .radio-btn {
 	display: block;
 	width: 100%;
+	max-width: 500px;
+	margin: 0 auto 0 auto;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/question/question.component.scss]: Same as above

File: exalearn-ui/src/app/components/speaking/speaking.component.scss [hunk 16]
```diff
@@ -1,18 +1,28 @@
+.spinnerPage {
+	margin: 25% auto 25% auto;
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/speaking/speaking.component.scss]: Same as above

File: exalearn-ui/src/app/components/data-table/data-table.component.ts (hunk 2 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/data-table/data-table.component.ts:21]: I don't know in full, it's just that the date is then distributed to all tables, and already inside the table, the type is assigned

File: exalearn-ui/src/app/components/timer/timer.component.ts [hunk 17]
```diff
@@ -18,40 +19,64 @@ export class TimerComponent implements OnInit, OnChanges, AfterViewInit {
...
-			this.time = this.timerService.displayTimeLeft(timerInterval, this.time.mins, this.time.secs);
-		}, 1000);
+	ngOnDestroy(): void {
+		this.stopTimerSubscribe();
+		this.timerService.pauseTimer();
+	}
+
+	testTimerSubscribe(): void {
+		this.testTimerSubscriber = this.timerService.testTimerObservable.subscribe((value) => {
+			if (value != undefined) {
```
Comment (YanaBr) [inline exalearn-ui/src/app/components/timer/timer.component.ts]: Just `if(value)`

File: exalearn-ui/src/app/interfaces/interfaces.ts [hunk 18]
```diff
@@ -11,9 +11,29 @@ export interface RedirectBtn {
 export interface Question {
 	id: number;
 	index: number | undefined;
+	question: object[];
+	answers?: any;
```
Comment (YanaBr) [inline exalearn-ui/src/app/interfaces/interfaces.ts]: Can we reolace this any?

File: exalearn-ui/src/app/interfaces/interfaces.ts [hunk 19]
```diff
@@ -11,9 +11,29 @@ export interface RedirectBtn {
...
-	questionText: string;
-	choices?: string[];
+	isCorrect: boolean;
+}
+
+export interface testAnswer {
+	id?: number;
+	passedTestId?: number;
+	questionId?: number;
+	answer?: any;
```
Comment (YanaBr) [inline exalearn-ui/src/app/interfaces/interfaces.ts]: Same

File: exalearn-ui/src/app/pages/test-page/test-page.component.ts [hunk 20]
```diff
@@ -1,44 +1,65 @@
...
 })
 export class TestPageComponent implements OnInit {
+	public testQuestions: Question[] = [];
+	public testQuestionsAudio: Question[] = [];
+	public textTopic: Question[] = [];
+	public testPassedId: number;
+	public innerText = 'TIME LEFT';
+	public isDataAvailable = false;
+	public startModal: any = StartTestModalComponent;
+	public test = this.submit.getTest();
```
Comment (YanaBr) [inline exalearn-ui/src/app/pages/test-page/test-page.component.ts]: Need to be placed on Init

File: exalearn-ui/src/app/pages/test-page/test-page.component.ts [hunk 21]
```diff
@@ -1,44 +1,65 @@
...
+		private submit: SubmitTestService
 	) {}
 
-	async ngOnInit() {
-		//TODO: values of levelType must be dynamic
-		this.generatedQuestions = await this.apiService
-			.getRequest(API_PATH.TEST, { levelType: 1 })
-			.then((response) => response);
+	ngOnInit() {
+		if (this.test === undefined) {
```
Comment (YanaBr) [inline exalearn-ui/src/app/pages/test-page/test-page.component.ts]: If(this.test)

**Review by HrebenYury — APPROVED** [review]

File: exalearn-ui/src/app/components/audition/audition.component.scss (hunk 8 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/audition/audition.component.scss]: Ok

File: exalearn-ui/src/app/components/audition/audition.component.ts (hunk 9 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/audition/audition.component.ts]: ok

File: exalearn-ui/src/app/components/essay-part/essay-part.component.scss (hunk 10 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/essay-part/essay-part.component.scss]: ok

File: exalearn-ui/src/app/components/essay-part/essay-part.component.scss (hunk 11 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/essay-part/essay-part.component.scss:34]: Angular material issues

File: exalearn-ui/src/app/components/finish-test-modal/finish-test-modal.component.scss (hunk 12 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/finish-test-modal/finish-test-modal.component.scss]: ok

File: exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss (hunk 14 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/leave-test-modal/leave-test-modal.component.scss]: ok

File: exalearn-ui/src/app/components/question/question.component.scss (hunk 15 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/question/question.component.scss]: ok

File: exalearn-ui/src/app/components/speaking/speaking.component.scss (hunk 16 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/speaking/speaking.component.scss]: ok

File: exalearn-ui/src/app/components/timer/timer.component.ts (hunk 17 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/components/timer/timer.component.ts]: ok

File: exalearn-ui/src/app/interfaces/interfaces.ts (hunk 18 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/interfaces/interfaces.ts]: ok

File: exalearn-ui/src/app/interfaces/interfaces.ts (hunk 19 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/interfaces/interfaces.ts]: ok

File: exalearn-ui/src/app/pages/test-page/test-page.component.ts (hunk 20 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/pages/test-page/test-page.component.ts]: ok

File: exalearn-ui/src/app/pages/test-page/test-page.component.ts (hunk 21 above)
Comment (GeranSkot) [reply exalearn-ui/src/app/pages/test-page/test-page.component.ts]: !test

**Review by YanaBr — APPROVED** [review]

**Review by HenadzV — APPROVED** [review]

**Review by vstrabko — APPROVED** [review]
Nice work!

**Pull request merged**